File: evd_authoring_app/src/model/evdScript/program/skills/simplePickAndPlace.js
import { MoveTrajectory } from '../primitives';
import { CloseGripper } from './closeGripper';
import { OpenGripper } from './openGripper';
import { Skill } from '../skill';
import { NodeParser } from '../../utilityFunctions';
import Blockly from 'blockly';


export class SimplePickAndPlace extends Skill {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'simple-pick-and-place';
    }

    static fullTypeString() {
        return Skill.fullTypeString() + SimplePickAndPlace.typeString();
    }

    constructor(startLocUuid=null, pickLocUuid=null, placeLocUuid=null, thingUuid=null, 
                type='', name='', uuid=null, parent=null, appendType=true, primitives=null) 
    {

        if (primitives === null) {
            primitives = [
                MoveTrajectory(startLocUuid,pickLocUuid),
                CloseGripper(thingUuid),
                MoveTrajectory(pickLocUuid,placeLocUuid),
                OpenGripper(thingUuid)
            ];
        }

        super(
            primitives,
            (appendType) ? 'simple-pick-and-place.'+type : type,
            name,
            uuid,
            parent,
            appendType
        );
    }

    static fromDict(dct) {
        return new SimplePickAndPlace(
            null,
            null,
            null,
            null,
            dct.type,
            dct.name,
            dct.uuid,
            null,
            false,
            dct.primitives.map(p => NodeParser(p))
        );
    }

    static BlocklyToolbox() {
        return { type: 'simple_pick_and_place' };
    }

    static BlocklyBlock() {
        return { 
            name: 'Simple Pick And Place',
            category: 'Skills',
            block: {
                init: function() {
                    this.appendDummyInput()
                        .appendField("Simple Pick and Place");
                    this.appendValueInput("thing")
                        .appendField("Thing");
                    this.appendValueInput("start-location")
                        .appendField("Start Location");
                    this.appendValueInput("pick-location")
                        .appendField("Pick Location");
                    this.appendValueInput("place-location")
                        .appendField("Place Location");
                    this.setPreviousStatement(true, null);
                    this.setNextStatement(true, null);
                    this.setColour(210);
                    this.setTooltip("simple-pick-and-place");
                    this.setHelpUrl("simple-pick-and-place");
                }
            },
            generator: (block) => {
                return `console.log('simple_pick_and_place')`;
            }
        };
    }
}

File: evd_ros_backend/evd_ros_core/src/evd_script/program/skills/simple_pick_and_place.py
'''
Pick-and-place is a very common activity for collaborative robots.

This is a simplified (non-conditional/error-handling) implementation that
commands an end-effector to move to a target object, grasp it, move to final
location, and then release it.
'''

from ..skill import Skill
from ..primitives import MoveTrajectory
from .open_gripper import OpenGripper
from .close_gripper import CloseGripper


class SimplePickAndPlace(Skill):

    '''
    Data structure methods
    '''

    @classmethod
    def type_string(cls, trailing_delim=True):
        return 'simple-pick-and-place' + ('.' if trailing_delim else '')

    @classmethod
    def full_type_string(cls):
        return Skill.full_type_string() + cls.type_string()

    def __init__(self, startLocUuid=None, pickLocUuid=None, placeLocUuid=None, thing_uuid=None,
                 type='', name='', uuid=None, parent=None, append_type=True, primitives=None, editable=True, deleteable=True):

        if primitives == None:
            primitives = [
                MoveTrajectory(startLocUuid,pickLocUuid, editable=editable, deleteable=deleteable),
                CloseGripper(thing_uuid=thing_uuid, editable=editable, deleteable=deleteable),
                MoveTrajectory(pickLocUuid,placeLocUuid, editable=editable, deleteable=deleteable),
                OpenGripper(thing_uuid=thing_uuid, editable=editable, deleteable=deleteable)
            ]

        super(SimplePickAndPlace,self).__init__(
            type=SimplePickAndPlace.type_string() + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type,
            primitives=primitives,
            editable=editable,
            deleteable=deleteable)
